app/models/fact_value.rb

```diff
@@ -81,13 +81,10 @@ def self.count_each(fact, options = {})
     output
   end
 
-  def self.build_facts_hash(facts)
-    hosts = Host.where(:id => facts.group_by(&:host_id).keys).all
-
-    hash = {}
-    facts.each do |fact|
-      hash[hosts.detect {|h| h.id == fact.host_id}.to_s] ||= {}
-      hash[hosts.detect {|h| h.id == fact.host_id}.to_s].update({fact.name.to_s => fact.value})
+  def self.build_facts_hash(values, host_id = nil)
+    hash  = values.group_by(&:host_name).transform_values!{|val| val.map{|v| [v.fact_name_name, v.value]}.to_h}
+    if host_id
+      hash  = hash.merge! hash.values[0]
```

Comment:
This line causes `no implicit conversion of nil into Hash` error, if there is no host present.